### DISCUSSION TITLE
Can timidity41 handle SoundFont larger than 2GB?

### DISCUSSION THREAD

**Discussion by knoike** [starg2/timidity41#4] [General]
Can timidity41 handle SoundFont larger than 2GB?
I want to to use HiDef SoundFont with timidity.
The file size of it is approximately 4GB.
I tried to use it in TiMidity++ on 64bit Linux, but some errors occurred and it could not be loaded.
I think that this was cause by the file size was too large.
(Is it the singed 32bit variable limitation?)

So I asked you this question.
Can timidity41 handle SoundFont larger than 2GB?
If the answer is Yes, I use timidity41 instead of TiMidity++.

Error msgs:
Trying to open /usr/local/share/timidity/HiDef/HiDef_20190525.sf2
INFO:
 ifil:
  version 2, minor 4
 isng:
  E-mu 10K1
 INAM:
  name 4GBC&C50_tmp.sf2
 ICRD:
  2/29/2016 2:48:21 AM
 IENG:
  SynthFont Viena
 IPRD:
  SynthFont Viena
 ICOP:
  SynthFont Viena
 ISFT:
  Polyphone
sdta:
/usr/local/share/timidity/HiDef/HiDef_20190525.sf2: *** illegal id in level 0: smpl -1471800400
/usr/local/share/timidity/HiDef/HiDef_20190525.sf2: *** illegal bags / gens

HiDef SoundFont:
https://musical-artifacts.com/artifacts/2525

**Comment by knoike** [comment]
ところで，ここで会話するときは，日本語で書いてもよいでしょうか．

**Comment by starg2** [comment]
日本語で大丈夫ですよ

結論から言うと、timidity41 は 2 GB 以上のサウンドフォントに対応しています

2 GB 以上のファイルに対応するためには、ファイルのシークが 64 ビットに対応している必要があり、その判定は `timidity/sysdep.h` で行っています。今どきの環境であればほぼ問題ないと思います

https://github.com/starg2/timidity41/blob/ca98f20e9edb147946b27b0dfb09d44533442c46/timidity/sysdep.h#L289-L336

ただし、本家 TiMidity++ には 64 ビット対応漏れがあり、 d2264d7f5f31da7422592c468029dbb32e2e9c64 の修正が必要ですが

**Reply by knoike** [reply]
おぉっ，すばらしい情報をありがとうございます．
ということは，

A/ timidity41 を Linux 上で動かす
B/ TiMidity++ に https://github.com/starg2/timidity41/commit/d2264d7f5f31da7422592c468029dbb32e2e9c64 を適用する

のどちらかで，Linux 上でも 2GB 超えサウンドフォントを扱えそうですね．

「The "make" command infinitely recurses, forever re-creating Makefile, and the product is never built」の
https://github.com/starg2/timidity41/discussions/3#discussioncomment-7351998
によると，timidity41 を Linux 上でビルドするときに，いまある configure はうまく動かないようですので，
CMakeLists.txt を直すか，autotools で configure を生成しなおすかになりますね．
時間を作って試してみたいと思います．ありがとうございました．
